File: 2022/OpenCV4 tutorial/QRCode/qrcode_decode.cpp
#include <opencv2/opencv.hpp>
using namespace cv;
void display(Mat &im, Mat &bbox)
{
  int n = bbox.rows;
  for(int i = 0 ; i < n ; i++)
  {
    line(im, Point2i(bbox.at<float>(i,0),bbox.at<float>(i,1)), Point2i(bbox.at<float>((i+1) % n,0), bbox.at<float>((i+1) % n,1)), Scalar(255,0,0), 3);
  }
  
  imshow("result", im);
}

int main(int argc, char* argv[])
{
    // Read image
    Mat inputImage;
    if(argc>1)
        inputImage = imread(argv[1]);
    else
        inputImage = imread("encode_data.jpg");

    QRCodeDetector qrDecoder;

    Mat bbox, rectifiedImage;

    std::string data = qrDecoder.detectAndDecode(inputImage, bbox, rectifiedImage);
    if(data.length()>0)
    {
        std::cout << "Decoded Data : " << data << std::endl;

        display(inputImage, bbox);
        rectifiedImage.convertTo(rectifiedImage, CV_8UC3);
        namedWindow("Rectified QRCode", WINDOW_NORMAL); // could be resized window
        imshow("Rectified QRCode", rectifiedImage);

        waitKey(0);
    }
    else
        std::cout << "QR Code not detected" << std::endl;
}

File: 2022/OpenCV4 tutorial/QRCode/qrcode_decode.cs
using System;
using System.IO;
using OpenCvSharp; //导入OpenCV4

namespace QRCode
{
    
    class Program
    {
        
        static void Main(string[] args)
        {
            DirectoryInfo rootDir = Directory.GetParent(Environment.CurrentDirectory);
            string root = rootDir.Parent.Parent.FullName;
            System.IO.Directory.SetCurrentDirectory(root);

            Mat inputImage = Cv2.ImRead("encode_data.jpg");
            
            if (!inputImage.Empty())
            {
                QRCodeDetector qrDecoder = new QRCodeDetector();

                Mat rectifiedImage = new Mat();

                Point2f[] bbox;

                string data = qrDecoder.DetectAndDecode(inputImage, out bbox, rectifiedImage);
                if (data.Length > 0)
                {
                    Console.WriteLine( "Decoded Data : {0}", data);

                    rectifiedImage.ConvertTo(rectifiedImage, MatType.CV_8UC3);
                    Cv2.NamedWindow("Rectified QRCode", WindowFlags.Normal); // could be resized window
                    Cv2.ImShow("Rectified QRCode", rectifiedImage);

                    Cv2.WaitKey(0);
                }
            }

        }
    }
}